feat: schema ingestion block

aspicer opened the pull request:
## Problem
we want the ability to allow users to reject events with bad schemas to keep their data clean

## Changes
add an option to have ingestion analyze the presence of and type of required properties. if they aren't right, don't accept them

## How did you test this code?
wrote tests for it, verified it locally

aspicer committed:
schema ingestion

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

aspicer committed:
tests

aspicer committed:
fix tests

aspicer committed:
snapshot

aspicer committed:
merge

aspicer committed:
migration

aspicer committed:
fix migrations

aspicer committed:
one more

aspicer committed:
fe

aspicer committed:
required properties

aspicer committed:
merge

aspicer committed:
hogli?

aspicer committed:
snapshot

aspicer committed:
max

aspicer committed:
merge

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

aspicer committed:
new file

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
merge

github-actions[bot] committed:
test(backend): update query snapshots

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

robbie-c commented on plugin-server/src/ingestion/event-preprocessing/validate-event-schema.ts:
```diff
@@ -0,0 +1,191 @@
+
+        case 'Object':
+            // Accepts: objects and arrays
+            return typeof value === 'object'
```
`typeof value === 'object'` seems a little scary because so many things in JS will match this
```
typeof null
typeof (new Promise(() => {}))
typeof (new Date())
```

Something like [lodash](https://github.com/lodash/lodash/blob/4.17.21/lodash.js#L12104) checks the prototype, so we could do that, though I think it's more important to match the query-time behaviour

robbie-c commented on plugin-server/src/ingestion/event-preprocessing/validate-event-schema.ts:
```diff
@@ -0,0 +1,191 @@
+
+/**
+ * Checks if a value can be coerced to the given PostHog property type.
+ * This matches PostHog's query-time type coercion behavior.
```
nit: could link to where this lives in the code

robbie-c reviewed (approved):
LGTM with a couple nits but I defer to @PostHog/team-ingestion

aspicer committed:
merge

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
clickhouse datetime

aspicer committed:
improve comments

aspicer committed:
merge clickhouse-datetime-parser branch

aspicer committed:
validate

aspicer committed:
chore(migrations): add schema_enforcement_mode field to EventDefinition

github-actions[bot] committed:
test(storybook): update UI snapshots

aspicer committed:
Merge branch 'aspicer/ingestion_types_migration' into aspicer/ingestion_types

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types_migration

aspicer committed:
manager

aspicer committed:
erge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

github-actions[bot] committed:
test(storybook): update UI snapshots

aspicer committed:
correct migration

aspicer committed:
Merge branch 'aspicer/ingestion_types_migration' into aspicer/ingestion_types

aspicer committed:
update

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
fix test

aspicer committed:
fix

aspicer committed:
addd

aspicer committed:
migration

aspicer committed:
merge

github-actions[bot] committed:
test(storybook): update UI snapshots

aspicer committed:
merge

aspicer committed:
merge

aspicer committed:
fix(ci): fetch base branch before migration detection in Node.js CI

The migration detection step was failing silently because
`origin/master` (or the base branch) wasn't fetched in the shallow
clone. This caused the fast path to always be taken, restoring
the schema artifact from master even when the PR contains migrations.

Add a fetch step to retrieve the base branch before running git diff.

aspicer committed:
Merge branch 'fix/ci-nodejs-migration-detection' into aspicer/ingestion_types

aspicer committed:
fix(ci): use two-dot diff to avoid merge base requirement

The three-dot diff requires a merge base between refs, which doesn't
exist with shallow fetches. Two-dot diff compares the refs directly.

aspicer committed:
chore: merge CI fix for migration detection

aspicer committed:
merge

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

jose-sequeira reviewed (dismissed):
This PR should be broken down into smaller PRs, adding a migration while changing ingestion code is not a good pattern. 

While the migrations and django are guaranteed to be synced, since django is the one executing them, this isn't necessarily true across all nodejs services. I'd recommend making ingestion changes on a separate PR

jose-sequeira commented on nodejs/src/utils/event-schema-enforcement-manager.ts:
```diff
@@ -0,0 +1,147 @@
+        // team_id -> event_name -> property_name -> property_types[]
+        const teamEventProps: Record<string, Record<string, Record<string, Set<string>>>> = {}
+
+        for (const row of rows) {
```
Triple nested `for loop` that can be run for a large quantity of rows and runs for every event in case of a cold cache. 

This adds to the impact of a cold cache and cache misses

jose-sequeira commented on nodejs/src/ingestion/event-preprocessing/validate-event-schema.ts:
```diff
@@ -0,0 +1,188 @@
+    eventName: string,
+    enforcedSchemas: EventSchemaEnforcement[]
+): EventSchemaEnforcement | undefined {
+    return enforcedSchemas.find((schema) => schema.event_name === eventName)
```
`Array.find`'s complexity is O(n), this runs on the hot path for potentially large amount of event definitions per team. This suggests that we have possibly chosen a wrong format to store our enforcedSchemas and maybe should store them as a HashMap<string, EventSchemaEnforcement>

jose-sequeira commented on posthog/models/schema.py:
```diff
@@ -113,6 +118,11 @@ class EventSchema(UUIDTModel):
         related_name="event_schemas",
         related_query_name="event_schema",
     )
+    enforcement_mode = models.CharField(
```
We filter on this field but there's no associated index

aspicer committed:
testing

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

github-actions[bot] committed:
test(backend): update query snapshots

aspicer committed:
chore(schema): add partial index on enforcement_mode for reject queries

Adds a partial index on EventSchema.enforcement_mode filtered for 'reject'
values, which optimizes the query used during ingestion to fetch enforced
schemas.

aspicer committed:
merge: include enforcement_mode index from migration branch

aspicer committed:
map

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
feat(schema): move enforcement_mode from EventSchema to EventDefinition

enforcement_mode is always toggled for all groups on an event
simultaneously, so it belongs on EventDefinition rather than the
EventSchema join table. This simplifies the data model and reduces
the ingestion query from a 4-table to 3-table join.

aspicer committed:
Merge branch 'aspicer/ingestion_types_migration' into aspicer/ingestion_types

aspicer committed:
feat(schema): update API, frontend, and ingestion for enforcement_mode on EventDefinition

- Remove enforcement_mode from EventSchemaSerializer
- Add enforcement_mode to EventDefinitionSerializer and
  EnterpriseEventDefinitionSerializer
- Simplify ingestion query to 3-table join (skip SchemaPropertyGroup)
- Update frontend logic to read/write enforcement_mode on
  EventDefinition instead of EventSchema
- Update all tests

aspicer committed:
change table

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
perf

aspicer committed:
fix

aspicer committed:
graphite

aspicer committed:
merge

github-actions[bot] committed:
test(backend): update query snapshots

aspicer committed:
test snapshot

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
schema

aspicer committed:
max migration

github-actions[bot] committed:
test(backend): update query snapshots

aspicer committed:
okay

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

jose-sequeira commented on nodejs/src/utils/event-schema-enforcement-manager.ts:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        // Simple flat query - aggregation happens in JS
+        const queryResult = await this.postgres.query<RawSchemaPropertyRow>(
```
This seems reasonable for the time being. The only request is that you have this step behind a configuration flag, this way we can bypass the step in case it is bringing us performance issues

aspicer committed:
flag

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

github-actions[bot] committed:
test(backend): update query snapshots

aspicer committed:
type fix

aspicer committed:
Merge branch 'aspicer/ingestion_types' of github.com:PostHog/posthog into aspicer/ingestion_types

aspicer committed:
merge

aspicer committed:
merge

aspicer committed:
refactor typing errors

aspicer committed:
schemas

aspicer committed:
greptile

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types

aspicer committed:
await

aspicer committed:
selector fix

aspicer committed:
ui

aspicer committed:
Merge remote-tracking branch 'origin/master' into aspicer/ingestion_types

aspicer committed:
Merge branch 'master' into aspicer/ingestion_types
